prediction_market_agent/agents/blockchain_coding_agent/functions.py

```diff
@@ -75,62 +75,45 @@ def execute_read_function(
 
     Returns:
         Any: The result of calling the specified function on the smart contract.
 
     """
-
-    c = ContractOnGnosisChain(
-        abi=abi_field_validator(abi), address=Web3.to_checksum_address(contract_address)
-    )
-    return c.call(function_name, function_parameters)
-
-
-def register_all_functions(
-    caller_agent: ConversableAgent, executor_agent: ConversableAgent
-) -> None:
-    # Register the search tool.
-    register_function(
-        tavily_search,
-        caller=caller_agent,
-        executor=executor_agent,
-        name="search_tool",
-        description="Search the web for the given query",
+    rpc_endpoint = get_rpc_endpoint()
+    w3 = Web3(Web3.HTTPProvider(rpc_endpoint))
+    contract = w3.eth.contract(
+        address=Web3.to_checksum_address(contract_address), abi=abi_field_validator(abi)
     )
 
-    # Register the calculator function to the two agents.
-    register_function(
-        fetch_read_methods_from_blockscout,
-        caller=caller_agent,  # The assistant agent can suggest calls to the calculator.
-        executor=executor_agent,  # The user proxy agent can execute the calculator calls.
-        name="fetch_read_methods_from_blockscout",  # By default, the function name is used as the tool name.
-        description="Function for fetching the ABI from a verified smart contract on Gnosis Chain",
-        # A description of the tool.
-    )
+    output = contract.functions[function_name](
+        *parse_function_params(function_params)
+    ).call()  # type: ignore # TODO: Fix Mypy, as this works just OK.
+    return output
 
-    register_function(
-        get_rpc_endpoint,
-        caller=caller_agent,  # The assistant agent can suggest calls to the calculator.
-        executor=executor_agent,  # The user proxy agent can execute the calculator calls.
-        name="get_rpc_endpoint",  # By default, the function name is used as the tool name.
-        description="Returns the RPC endpoint to be used for interacting with Gnosis Chain when instantiating a provider",
-        # A description of the tool.
-    )
 
-    register_function(
-        execute_read_function,
-        caller=caller_agent,
-        executor=executor_agent,
-        name="execute_read_function",
-        description="Executes a function on a smart contract deployed on the Gnosis Chain and returns the result of the function execution.",
-    )
+##### Write
 
-    register_function(
-        checksum_address,
-        caller=caller_agent,
-        executor=executor_agent,
-        name="checksum_address",
-        description="Extracts the checksummed address of an address.",
+
+def get_private_key() -> PrivateKey:
+    return APIKeys().bet_from_private_key
+
+
+def execute_write_function(
+    contract_address: str,
+    abi: str,
+    function_name: str,
+    function_params: t.List[t.Any],
+) -> TxReceipt:
+    # ToDo - Do not use PMAT if possible.
```

Comment:
Why not? If there is something implemented in PMAT already, I don't see reason why not just wrap it into autogen-compatible functions?

Reply:
See https://github.com/gnosis/prediction-market-agent/pull/566/files#r1872193547
Closing this since same idea from other comment